File: chalawa_package_frontend/frontend/lib/src/decryption.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:pointycastle/export.dart';
import 'types.dart';

/// Decrypts encrypted text using AES-256-GCM with password-derived key
dynamic decrypt(DecryptionInput input) {
  try {
    // Parse encrypted text format: encryptedData:iv:authTag
    final parts = input.encryptedText.split(':');
    if (parts.length != 3) {
      throw ArgumentError('Invalid encrypted text format');
    }

    final encryptedData = base64.decode(parts[0]);
    final iv = base64.decode(parts[1]);
    final authTag = base64.decode(parts[2]);

    // Create key from password using SHA-512 hash (same as Node.js implementation)
    final hash = sha512.convert(utf8.encode(input.password));
    final keyString = hash.toString();
    final key = utf8.encode(keyString.substring(0, 32));

    // Combine encrypted data and auth tag for GCM
    final combinedData = Uint8List(encryptedData.length + authTag.length);
    combinedData.setAll(0, encryptedData);
    combinedData.setAll(encryptedData.length, authTag);

    // Initialize AES-256-GCM cipher
    final cipher = GCMBlockCipher(AESEngine());
    final keyParam = KeyParameter(Uint8List.fromList(key));
    final params = AEADParameters(keyParam, 128, iv, Uint8List(0)); // 128-bit auth tag
    
    cipher.init(false, params); // false = decrypt

    // Decrypt
    final decryptedBytes = cipher.process(combinedData);
    
    // Convert back to string and parse JSON
    final decryptedString = utf8.decode(decryptedBytes);
    return jsonDecode(decryptedString);
  } catch (e) {
    throw Exception('Decryption failed: $e');
  }
}

File: chalawa_package_frontend/backend/src/decryption.ts
import { createCipheriv, createDecipheriv, createHash, randomBytes } from "crypto";
import { DecryptionInput } from "./types";

export function decrypt(
    {
        encryptedText,
        password
    }: DecryptionInput
) {
    let m = createHash('sha512');
    let dataKey = m.update(password, 'utf-8');
    let genHash = dataKey.digest('hex');
    let key = genHash.substring(0, 32);
    let result = encryptedText.split(":");
    if (result.length !== 3) {
        throw new Error('Invalid encrypted text format');
    }
    let iv = Buffer.from(result[1], 'base64');
    let authTag = Buffer.from(result[2], 'base64');
    let decipher = createDecipheriv('aes-256-gcm', Buffer.from(key), iv);
    decipher.setAuthTag(authTag);
    let decrypted = decipher.update(result[0], 'base64', 'utf-8');
    decrypted += decipher.final('utf-8');
    return JSON.parse(decrypted);
}
